### ISSUE TITLE
Holy bolts too often fail to hit the target

### ISSUE THREAD

**Issue by Chance4us** [diasurgical/devilutionX#2890]
I've noticed that although holy bolts are shot straight ahead at the skeletons,
an unusual number of times the holy bolts fly through without hitting the skeletons.
Unfortunately it could not be narrowed down so far and this issue serves as a reminder to complete further.

EDIT:
It happens very often when the skeleton is very close to the hero.

**Comment by StephenCWills** [comment]
It could be a consequence of #2831. It could be useful to know what level of Holy Bolt you are testing, and also some idea of what level of Holy Bolt you typically use.

**Comment by FitzRoyX** [comment]
I tried to reproduce this and couldn't

**Comment by Chance4us** [comment]
I tried a game as a low leveled rogue in normal difficulty.
The way of playing is not to clear the levels completely, but to descend as quickly as possible.

As soon as I got the holy bolt spell (level 1) I had the magic at 20 and very few mana points.
So that it was very critical to miss the target and the game turned out to be very difficult.

Since the "chance to hit" rate at rogue is relatively high right from the beginning, the missing to hit the target felt anomalous.

Unfortunately I need some time to make some comparisons because it is a complex problem.

**Comment by Chance4us** [comment]
> It could be a consequence of #2831.

I will leave a comment there so that we do not deal with different topics here.

**Comment by Chance4us** [comment]
I have now compared vanilla with master. There were 50 shots and the hits and misses were counted.
Vanilla hit rate around 80%, approximately 40 hits and 10 misses.

https://user-images.githubusercontent.com/45902449/133887170-0e4bb585-7ab3-449b-a829-c2fec796a010.mp4

In the same way, I counted 27 hits and 23 misses in master.
You can use the attached savegame to test it.
[hits.zip](https://github.com/diasurgical/devilutionX/files/7189807/hits.zip)

**Comment by Chance4us** [comment]
After bisecting it was discovered that the issue appeared since commit 836bb92e8ed6b602b980ee2393dcf93d66524f6f.

**Comment by StephenCWills** [comment]
Based on your numbers, here is a chance to miss calculation.

1 collision per tile: `23 / 50 = 0.46`
2 collisions per tile: `0.46^2 = 0.212`
3 collisions per tile: `0.46^3 = 0.097`

Your 80% calculation in vanilla matches up quite well with 2 collisions per tile so this is all coming together. Unfortunately, a 30% drop in chance to hit is indeed pretty significant, and it's difficult to reconcile this with how OP certain high level spells with fewer hit checks already are. I'm honestly not sure what to do here, but I'll think on it.

**Comment by NiteKat** [comment]
> I tried a game as a low leveled rogue in normal difficulty.
> The way of playing is not to clear the levels completely, but to descend as quickly as possible.
> 
> As soon as I got the holy bolt spell (level 1) I had the magic at 20 and very few mana points.
> So that it was very critical to miss the target and the game turned out to be very difficult.
> 
> Since the "chance to hit" rate at rogue is relatively high right from the beginning, the missing to hit the target felt anomalous.
> 
> Unfortunately I need some time to make some comparisons because it is a complex problem.

Isn't chance to hit for spells based on the character's magic stat, so 20 magic would give a pretty low chance for your spells to land? I thought spells didn't care about the To Hit stat.

**Comment by StephenCWills** [comment]
> Isn't chance to hit for spells based on the character's magic stat, so 20 magic would give a pretty low chance for your spells to land? I thought spells didn't care about the To Hit stat.

That does appear to be the case.

`PercentChanceToHit = 50 + Magic + ClassBonus - (2*MonsterLevel) - Distance`

`ClassBonus` is 20 for Sorcerer and 10 for Bard.

**Comment by AJenbo** [comment]
> I have now compared vanilla with master. There were 50 shots and the hits and misses were counted.
> Vanilla hit rate is around 80%, approximately 40 hits and 10 misses.
> vanilla.mp4
> 
> In the same way, I counted 27 hits and 23 misses in master.
> You can use the attached savegame to test it.
> [hits.zip](https://github.com/diasurgical/devilutionX/files/7189807/hits.zip)

This might not be the best way to test the to hit change, since loading the save game will revert the RNG to the same state, so thing will be heavily biased based on the exact timing of your moves.

> Isn't chance to hit for spells based on the character's magic stat
Correct, this is the formula:
```cpp
		int chance = magic + 50;
		if (HeroClass::Sorcerer)
			chance += 20;
		else if (HeroClass::Bard)
			chance += 10;
		chance -= monster.mLevel * 2;
```
So for a Rouge with 20 magic fighting a Burning dead you should see a 56% chance to hit. Which actually fit very accurately with @Chance4us's test where he had a 54% hit rate.

**Comment by AJenbo** [comment]
> `PercentChanceToHit = 50 + Magic + ClassBonus - (2*MonsterLevel) - Distance`

On a side note, I just noticed that Distance is only increased for arrows (and HorkSpawn for no good reason), making it always 0 for spells.

**Comment by Chance4us** [comment]
@AJenbo Do you like to play in German translation?

**Comment by AJenbo** [comment]
> @AJenbo Do you like to play in German translation?

I have it loaded atm as I'm working on improving the translation support. But I do enjoy how I'm able to slowly understand more and more of it.

**Comment by julealgon** [comment]
I love seeing this issue being opened. It means the fix to missile collisions worked perfectly :)

@soundchaser83 FYI

> Unfortunately, a 30% drop in chance to hit is indeed pretty significant, and it's difficult to reconcile this with how OP certain high level spells with fewer hit checks already are. I'm honestly not sure what to do here, but I'll think on it.

I don't think we should do anything. That's my vote, at least.

We should close this.

**Comment by StephenCWills** [comment]
> I don't think we should do anything. That's my vote, at least.
>
> We should close this.

I would agree, but I think it's worth playtesting Sorcerer to see how much more difficult the early levels have become.

**Comment by julealgon** [comment]
> I would agree, but I think it's worth playtesting Sorcerer to see how much more difficult the early levels have become.

My prediction is that it will be overall _slightly_ harder than before. Yes it will be harder to hit in situations where hit was nearly guaranteed before, but at the same time it should be much easier to hit in _other situations where it was nearly impossible to hit before_, like moving targets and monsters in weird angles.

To me, then change makes the game far more intuitive and consistent with the new behaviors.

**Comment by SoundChaser83** [comment]
It seems like this is just #2831 working as intended; since that change, we should expect the master version to yield lower effective hit chance on slower spells like Holy Bolt compared to vanilla. We're probably all used to the feel of Holy Bolt in vanilla where the game likely was giving the missile multiple checks to roll a hit as it passed over the monster, but now that shouldn't be happening anymore.

**Comment by StephenCWills** [comment]
> My prediction is that it will be overall _slightly_ harder than before.

I did a bit of playtesting just now on both mouse/keyboard and controller. It didn't even feel slightly harder than what I'm used to, to be honest. I can't safely lure the enemies to a neighboring square to get a near-guaranteed hit, so I was getting hit a lot more often at first. I had to overcome some of my predispositions, but I found that if I line up on the diagonals like usual, I can still count on my spells to hit enemies on approach. Even if I don't line up on the diagonals, I have a marginally easier time hitting enemies at odd angles so I didn't find myself wasting a lot of mana or staff charges. Ultimately, I found I was able to play a bit more trigger happy and still be successful, which might actually make Sorcerer feel easier to newcomers.
